injector/injector.go

```diff
@@ -0,0 +1,302 @@
+	companionClickThroughEndTag    = "]]></CompanionClickThrough>"
+	tracking                       = "tracking"
+	companionclickthrough          = "companionclickthrough"
+	nonlinearclicktracking         = "nonlinearclicktracking"
+	impression                     = "impression"
+	err                            = "error"
+	clicktracking                  = "clicktracking"
+)
+
+type Injector interface {
+	InjectTracker(vastXML string, NURL string) string
+}
+
+type VASTEvents struct {
+	Errors                  []string
+	Impressions             []string
+	VideoClicks             []string
+	NonLinearClickTracking  []string
+	CompanionClickThrough   []string
+	LinearTrackingEvents    map[string][]string
+	NonLinearTrackingEvents map[string][]string
+	CompanionTrackingEvents map[string][]string
+}
+
+type TrackerInjector struct {
+	replacer macros.Replacer
+	events   VASTEvents
+	me       metrics.MetricsEngine
+	provider *macros.MacroProvider
+}
+
+var trimRunes = "\t\r\b\n "
+
+func NewTrackerInjector(replacer macros.Replacer, provider *macros.MacroProvider, events VASTEvents) *TrackerInjector {
+	return &TrackerInjector{
+		replacer: replacer,
+		provider: provider,
+		events:   events,
+	}
+}
+
+func (ti *TrackerInjector) InjectTracker(vastXML string, NURL string) string {
+	if vastXML == "" && NURL == "" {
+		// TODO Log a adapter.<bidder-name>.requests.badserverresponse
+		return vastXML
+	}
+
+	if vastXML == "" {
+		return fmt.Sprintf(emptyAdmResponse, NURL)
+	}
+
+	var outputXML strings.Builder
+	encoder := xml.NewEncoder(&outputXML)
+
+	injectTracker := false
+	injectVideoClicks := false
+	inlineWrapperTagFound := false
+	wrapperTagFound := false
+	impressionTagFound := false
+	errorTagFound := false
+	creativeId := ""
+	isCreative := false
+	companionTagFound := false
+	nonLinearTagFound := false
+
+	reader := strings.NewReader(vastXML)
+	decoder := xml.NewDecoder(reader)
+
+	for {
+		t, err := decoder.RawToken()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return ""
+		}
+
+		switch tt := t.(type) {
+		case xml.StartElement:
+			switch tt.Name.Local {
+			case "Wrapper":
+				wrapperTagFound = true
+			case "Creative":
+				isCreative = true
+				for _, attr := range tt.Attr {
+					if strings.ToLower(attr.Name.Local) == "adid" {
+						creativeId = attr.Value
+					}
+				}
+			case "Linear":
+				injectVideoClicks = true
+				injectTracker = true
+			case "VideoClicks":
+				injectVideoClicks = false
+				encoder.Flush()
+				encoder.EncodeToken(tt)
+				encoder.Flush()
+				ti.addClickTrackingEvent(&outputXML, creativeId, false)
+
+				continue
+			case "NonLinearAds":
+				injectTracker = true
+			case "TrackingEvents":
+				if isCreative {
+					injectTracker = false
+					encoder.Flush()
+					encoder.EncodeToken(tt)
+					encoder.Flush()
+					ti.addTrackingEvent(&outputXML, creativeId, false)
+					continue
+				}
+			}
+
+		case xml.EndElement:
+			switch tt.Name.Local {
```

Comment:
And maybe we could do the same here with `xml.EndElement`.
```
148         case xml.EndElement:
    +           addXmlEndElement(tt.Name.Local)
149 -           switch tt.Name.Local {
150 -           case "Impression":
151 -               encoder.Flush()
152 -               encoder.EncodeToken(tt)
153 -               encoder.Flush()
154 -               if !impressionTagFound {
155 -                   ti.addImpressionTrackingEvent(&outputXML)
156 -                   impressionTagFound = true
157 -               }
158 -               continue
159 -           case "Error":
160 -               encoder.Flush()
161 -               encoder.EncodeToken(tt)
162 -               encoder.Flush()
163 -               if !errorTagFound {
164 -                   ti.addErrorTrackingEvent(&outputXML)
165 -                   errorTagFound = true
166 -               }
167 -               continue
168 -           case "NonLinearAds":
169 -               if injectTracker {
170 -                   injectTracker = false
171 -                   encoder.Flush()
172 -                   ti.addTrackingEvent(&outputXML, creativeId, true)
173 -                   if !nonLinearTagFound && wrapperTagFound {
174 -                       ti.addNonLinearClickTrackingEvent(&outputXML, creativeId, true)
175 -                   }
176 -                   encoder.EncodeToken(tt)
177 -               }
178 -           case "Linear":
179 -               if injectVideoClicks {
180 -                   injectVideoClicks = false
181 -                   encoder.Flush()
182 -                   ti.addClickTrackingEvent(&outputXML, creativeId, true)
183 -               }
184 -               if injectTracker {
185 -                   injectTracker = false
186 -                   encoder.Flush()
187 -                   ti.addTrackingEvent(&outputXML, creativeId, true)
188 -               }
189 -               encoder.EncodeToken(tt)
190 -               continue
191 -           case "InLine", "Wrapper":
192 -               wrapperTagFound = false
193 -               inlineWrapperTagFound = true
194 -               encoder.Flush()
195 -               if !impressionTagFound {
196 -                   ti.addImpressionTrackingEvent(&outputXML)
197 -               }
198 -               impressionTagFound = false
199 -               if !errorTagFound {
200 -                   ti.addErrorTrackingEvent(&outputXML)
201 -               }
202 -               errorTagFound = false
203 -               encoder.EncodeToken(tt)
204 -           case "NonLinear":
205 -               encoder.Flush()
206 -               ti.addNonLinearClickTrackingEvent(&outputXML, creativeId, false)
207 -               nonLinearTagFound = true
208 -               encoder.EncodeToken(tt)
209 -           case "Companion":
210 -               companionTagFound = true
211 -               encoder.Flush()
212 -               ti.addCompanionClickThroughEvent(&outputXML, creativeId, false)
213 -               encoder.EncodeToken(tt)
214 -           case "Creative":
215 -               isCreative = false
216 -           case "CompanionAds":
217 -               if !companionTagFound && wrapperTagFound {
218 -                   encoder.Flush()
219 -                   ti.addCompanionClickThroughEvent(&outputXML, creativeId, true)
220 -               }
221 -           }
222         case xml.CharData:
223             tt2 := strings.Trim(string(tt), trimRunes)
224             if len(tt2) != 0 {
225                 encoder.Flush()
injector/injector.go
```

Reply:
Done